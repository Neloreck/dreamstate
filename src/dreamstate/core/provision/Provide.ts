import { default as hoistNonReactStatics } from "hoist-non-react-statics";
import { ComponentType } from "react";

import { createManagersObserver } from "@/dreamstate/core/observing/createManagersObserver";
import { createClassWrapDecorator } from "@/dreamstate/polyfills/createClassWrapDecorator";
import { TDreamstateService } from "@/dreamstate/types";

/**
 * Decorator factory.
 * Provide context from context managers.
 * Observes changes and uses default react Providers for data flow.
 */
export function Provide<T extends ComponentType>(
  sources: Array<TDreamstateService<any>>
): ClassDecorator {
  return createClassWrapDecorator(function(targetClass: T): T {
    return hoistNonReactStatics(createManagersObserver(targetClass, sources), targetClass);
  });
}
