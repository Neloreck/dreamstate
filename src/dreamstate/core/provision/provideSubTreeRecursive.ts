import { createElement, ReactElement } from "react";

import { CONTEXT_STATES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function provideSubTreeRecursive(
  bottom: ReactElement,
  sources: Array<TAnyContextManagerConstructor>,
  current: number
): ReactElement {
  return current >= sources.length
    ? bottom
    : createElement(
      sources[current].REACT_CONTEXT.Provider,
      { value: CONTEXT_STATES_REGISTRY.get(sources[current]) },
      provideSubTreeRecursive(bottom, sources, current + 1)
    );
}
