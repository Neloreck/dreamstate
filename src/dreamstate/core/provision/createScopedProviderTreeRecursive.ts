import { createElement, ReactNode } from "react";

import { ScopedProvider } from "@/dreamstate/core/provision/ScopedProvider";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function createScopedProviderTreeRecursive<T extends TAnyObject>(
  bottom: ReactNode = null,
  sources: Array<TAnyContextManagerConstructor>,
  current: number,
  initialState?: T
): ReactNode {
  return current >= sources.length
    ? bottom
    : createElement(
      ScopedProvider,
      { manager: sources[current], initialState },
      createScopedProviderTreeRecursive(bottom, sources, current + 1, initialState)
    );
}
