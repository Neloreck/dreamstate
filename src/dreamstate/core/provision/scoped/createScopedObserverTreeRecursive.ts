import { createElement, ReactNode } from "react";

import { ScopedObserver } from "@/dreamstate/core/provision/combined/ScopedObserver";
import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";
import { IProviderProps } from "@/dreamstate/types/provision";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function createScopedObserverTreeRecursive<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>,
  props: IProviderProps<T>,
  scope: IScopeContext,
  current: number = 0
): ReactNode {
  return current >= sources.length
    ? props.children
    : createElement(
      ScopedObserver,
      { manager: sources[current], initialState: props.initialState, dependencies: sources, scope },
      createScopedObserverTreeRecursive(sources, props, scope,current + 1)
    );
}
