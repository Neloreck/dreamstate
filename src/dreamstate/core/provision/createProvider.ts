import { FunctionComponent, ReactNode, useCallback, useState } from "react";

import { EMPTY_ARR } from "@/dreamstate/core/internals";
import { useHotObservers } from "@/dreamstate/core/observing/useHotObservers";
import { useStaticObservers } from "@/dreamstate/core/observing/useStaticObservers";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { ContextService } from "@/dreamstate/core/services/ContextService";
import {
  TAnyContextManagerConstructor,
  TAnyContextServiceConstructor,
  TAnyObject
} from "@/dreamstate/types";

export interface IProviderProps<T> {
  initialState?: T;
  partialHotReplacement?: boolean;
  children?: ReactNode;
}

/**
 * Create component for manual provision without HOC/Decorator-like api.
 * Useful if your root is functional component or you are using createComponent api without JSX.
 */
export function createProvider<
  T extends TAnyObject = TAnyObject
>(
  sources: Array<TAnyContextServiceConstructor>
): FunctionComponent<IProviderProps<T>> {
  if (!Array.isArray(sources)) {
    throw new TypeError(
      "Wrong providers parameter supplied. Only array of context services is acceptable."
    );
  }

  for (let it = 0; it < sources.length; it ++) {
    if (!sources[it] || !(sources[it].prototype instanceof ContextService)) {
      throw new TypeError("Only classes extending ContextService can be supplied for provision.");
    }
  }

  /**
   * Check only managers with required provision.
   * Do not include services for subTree rendering but add registering logic for services.
   */
  const managers: Array<TAnyContextManagerConstructor> = sources.filter(function(
    Service: TAnyContextServiceConstructor
  ) {
    return Service.prototype instanceof ContextManager;
  }) as Array<TAnyContextManagerConstructor>;

  /**
   * Create observer component that will handle observing.
   */
  function Observer(props: IProviderProps<T>): ReactNode {
    const [ , forceRender ] = useState({});
    const updateProviders = useCallback(function() {
      forceRender({});
    }, EMPTY_ARR);

    if (props.partialHotReplacement) {
      useHotObservers(sources, props.initialState, updateProviders);
    } else {
      useStaticObservers(sources, props.initialState, updateProviders);
    }

    return provideSubTreeRecursive(props.children, managers, 0);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextServiceConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
