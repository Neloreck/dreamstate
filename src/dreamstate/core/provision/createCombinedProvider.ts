import { FunctionComponent, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { useSourceObserving } from "@/dreamstate/core/provision/useSourceObserving";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Create provider that unifies all data sources and tries to re-render providers on every update.
 */
export function createCombinedProvider<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Observer(props: IProviderProps<T>): ReactNode {
    const registry: Map<TAnyContextManagerConstructor, TAnyObject> = useSourceObserving(sources, props);

    return provideSubTreeRecursive(props.children, sources, registry);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextManagerConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
