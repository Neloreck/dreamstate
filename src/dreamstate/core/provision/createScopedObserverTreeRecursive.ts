import { createElement, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { ScopedObserver } from "@/dreamstate/core/provision/ScopedObserver";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function createScopedObserverTreeRecursive<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>,
  props: IProviderProps<T>,
  current: number = 0
): ReactNode {
  return current >= sources.length
    ? props.children
    : createElement(
      ScopedObserver,
      { manager: sources[current], initialState: props.initialState, dependencies: sources },
      createScopedObserverTreeRecursive(sources, props, current + 1)
    );
}
