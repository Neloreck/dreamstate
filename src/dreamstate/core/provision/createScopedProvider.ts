import { FunctionComponent, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { createScopedObserverTreeRecursive } from "@/dreamstate/core/provision/createScopedObserverTreeRecursive";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Create provider that bundles all observers in a scoped nodes and
 * it does not affect all providers if only one was actually changed.
 */
export function createScopedProvider<T extends IProviderProps<any>>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Observer(props: T): ReactNode {
    return createScopedObserverTreeRecursive(sources, props);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextManagerConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
