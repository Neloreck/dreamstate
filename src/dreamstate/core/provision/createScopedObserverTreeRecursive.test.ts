import { shallow } from "enzyme";
import { createElement } from "react";

import { createScopedObserverTreeRecursive } from "@/dreamstate/core/provision/createScopedObserverTreeRecursive";
import { NestedContextManager, TestContextManager } from "@/fixtures";

describe("createScopedObserverTreeRecursive rendering", () => {
  it("Should match snapshot", () => {
    const recursive = shallow(
      createElement(
        "div",
        {},
        createScopedObserverTreeRecursive(
          [
            TestContextManager,
            NestedContextManager,
            TestContextManager
          ],
          { children: createElement("div", {}, "bottom") }
        )
      )
    );

    expect(recursive).toMatchSnapshot();
  });
});
