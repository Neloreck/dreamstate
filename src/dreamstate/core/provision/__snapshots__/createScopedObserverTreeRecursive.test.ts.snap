// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createScopedObserverTreeRecursive rendering Should match snapshot 1`] = `
<div>
  <ScopedObserver
    dependencies={
      Array [
        [Function],
        [Function],
        [Function],
      ]
    }
    manager={[Function]}
  >
    <ScopedObserver
      dependencies={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
      manager={[Function]}
    >
      <ScopedObserver
        dependencies={
          Array [
            [Function],
            [Function],
            [Function],
          ]
        }
        manager={[Function]}
      >
        <div>
          bottom
        </div>
      </ScopedObserver>
    </ScopedObserver>
  </ScopedObserver>
</div>
`;
